In [86]:
import pandas as pd
import numpy as np
from sklearn.metrics import confusion_matrix
import sklearn.metrics

### 2. Given the following confusion matrix, evaluate (by hand) the model's performance.

|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |


#### a. In the context of this problem, what is a false positive?

#### b. In the context of this problem, what is a false negative?

#### c. How would you describe this model?

In [87]:
r = {'pred_dog': 46, 'pred_cat':7}
df = pd.DataFrame(data=r, index=['actual_dog', 'actual_cat'], columns=['pred_dog','pred_cat'])
df.pred_dog['actual_cat'] = df.pred_dog['actual_cat'] - 33
df.pred_cat['actual_cat'] = df.pred_cat['actual_cat'] + 27
df

pred_dog  pred_cat
actual_dog        46         7
actual_cat        13        34

-Positive case: a dog

-Negative case: a cat

**Define the outcomes**

In [88]:
#True positive: Predict dog and actual dog
tp = df.pred_dog[0]

#True negative: Predict cat and actual cat
tn = df.pred_cat[1]

#False positive: Predict dog and actual cat
fp = df.pred_dog[1]

#False negative: Predict cat and actual dog
fn = df.pred_cat[0]

**Accuracy**: (TP + TN) / (TP + TN + FP + FN)

In [89]:
print(f'{round((tp + tn) / (tp + tn + fp + fn), 2)}%')

0.8%


**Precision**: TP / (TP + FP)

In [90]:
print(f'{round(tp/(tp+fp), 2)}%')

0.78%


**Recall**: TP / (TP + FN)

In [91]:
print(f'{round(tp/(tp+fn), 2)}%')

0.87%


This model is used to predict wether an animal is a cat or dog with a respectable success-rate.

### 3. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant. Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects. Use the predictions dataset and pandas to help answer the following questions:

#### a. An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

In [92]:
df = pd.read_csv('c3.csv')
df

actual     model1     model2     model3
0    No Defect  No Defect     Defect  No Defect
1    No Defect  No Defect     Defect     Defect
2    No Defect  No Defect     Defect  No Defect
3    No Defect     Defect     Defect     Defect
4    No Defect  No Defect     Defect  No Defect
..         ...        ...        ...        ...
195  No Defect  No Defect     Defect     Defect
196     Defect     Defect  No Defect  No Defect
197  No Defect  No Defect  No Defect  No Defect
198  No Defect  No Defect     Defect     Defect
199  No Defect  No Defect  No Defect     Defect

[200 rows x 4 columns]

**Cases**

-Positive case: Defect

-Negative case: No Defect

**Outcomes**

    True Posititve: Defective product marked defective
    
    True Negative: Working product marked not defective
    
    False Positive: Defective product marked not defective
    
    False Negative: Working product marked defective
    
    
**Model 1:**

In [93]:
# We first make a table for the first model and turn it into a cross table
df1 = df[['actual','model1']].copy()

df_cross1 = pd.crosstab(df1.model1, df1.actual)
df_cross1

actual     Defect  No Defect
model1                      
Defect          8          2
No Defect       8        182

Since we are looking to catch all defects (true positives and false negatives), we would find all defects and not just correctly predicted defects.

Therefore, the recall metric is the mnost appropriate.

In [94]:
df.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

In [95]:
# We add a column for the baseline prediction ('No Defect' since it is the most common outcome)
df1['baseline_pred'] = 'No Defect'

In [96]:
model_accuracy = (df1.actual == df1.model1).mean()
baseline_accuracy = (df1.actual == df1.baseline_pred).mean()

print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model accuracy: 95.00%
baseline accuracy: 92.00%


In [97]:
# Recall - calculated using subset
# Only looking at actual 'Defects' (i.e actual == Defect)
subset1 = df1[df1.actual == 'Defect']
subset1

actual     model1 baseline_pred
13   Defect  No Defect     No Defect
30   Defect     Defect     No Defect
65   Defect     Defect     No Defect
70   Defect     Defect     No Defect
74   Defect  No Defect     No Defect
87   Defect  No Defect     No Defect
118  Defect  No Defect     No Defect
135  Defect     Defect     No Defect
140  Defect  No Defect     No Defect
147  Defect     Defect     No Defect
163  Defect     Defect     No Defect
171  Defect  No Defect     No Defect
176  Defect  No Defect     No Defect
186  Defect  No Defect     No Defect
194  Defect     Defect     No Defect
196  Defect     Defect     No Defect

In [98]:
# calculate recall for model and baseline
model_recall = (subset1.actual == subset1.model1).mean()

print(f'model 1 recall: {model_recall:.2%}')

model 1 recall: 50.00%


We repeat the process for the next models:

**Model 2:**

In [99]:
df2 = df[['actual','model2']].copy()
df_cross2 = pd.crosstab(df2.model2, df2.actual)
df_cross2

actual     Defect  No Defect
model2                      
Defect          9         81
No Defect       7        103

In [100]:
model_accuracy = (df2.actual == df2.model2).mean()

print(f' model 2 accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

 model 2 accuracy: 56.00%
baseline accuracy: 92.00%


In [101]:
# Recall - calculate on subset of data
# Only looking at actual 'Defects' (i.e actual == Defect)
subset2 = df2[df2.actual == 'Defect']
subset2

actual     model2
13   Defect     Defect
30   Defect  No Defect
65   Defect     Defect
70   Defect     Defect
74   Defect  No Defect
87   Defect     Defect
118  Defect     Defect
135  Defect  No Defect
140  Defect     Defect
147  Defect  No Defect
163  Defect     Defect
171  Defect     Defect
176  Defect     Defect
186  Defect  No Defect
194  Defect  No Defect
196  Defect  No Defect

In [102]:
# calculate recall for model and baseline
model_recall = (subset2.actual == subset2.model2).mean()

print(f'model 2 recall: {model_recall:.2%}')

model 2 recall: 56.25%


**Model 3:**

In [103]:
df3 = df[['actual','model3']].copy()
df_cross3 = pd.crosstab(df3.model3, df3.actual)
df_cross3

actual     Defect  No Defect
model3                      
Defect         13         86
No Defect       3         98

In [104]:
model_accuracy = (df3.actual == df3.model3).mean()

print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model accuracy: 55.50%
baseline accuracy: 92.00%


In [105]:
# Recall - calculate on subset of data
# Only looking at actual 'Defects' (i.e actual == Defect)
subset3 = df3[df3.actual == 'Defect']
subset3

actual     model3
13   Defect     Defect
30   Defect     Defect
65   Defect     Defect
70   Defect     Defect
74   Defect     Defect
87   Defect     Defect
118  Defect  No Defect
135  Defect     Defect
140  Defect     Defect
147  Defect     Defect
163  Defect     Defect
171  Defect     Defect
176  Defect     Defect
186  Defect  No Defect
194  Defect     Defect
196  Defect  No Defect

In [106]:
# calculate recall for model and baseline
model_recall = (subset3.actual == subset3.model3).mean()

print(f'model 3 recall: {model_recall:.2%}')

model 3 recall: 81.25%


Since model 3 has the highest recall percentage of all models, model 3 would be the best model to use.

#### b. Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

Since we are trying to avoid false positives, the precision metric would be the best tool to use.

**Model 1:**

In [107]:
#We make a subset of observations with all predictions being positives
subset1 = df1[df1.model1 == 'Defect']
subset1

actual  model1 baseline_pred
3    No Defect  Defect     No Defect
30      Defect  Defect     No Defect
62   No Defect  Defect     No Defect
65      Defect  Defect     No Defect
70      Defect  Defect     No Defect
135     Defect  Defect     No Defect
147     Defect  Defect     No Defect
163     Defect  Defect     No Defect
194     Defect  Defect     No Defect
196     Defect  Defect     No Defect

In [108]:
model_precision = (subset1.actual == subset1.model1).mean()
print(f'model 1 precision: {model_precision:.2%}')

model 1 precision: 80.00%


**Model 2:**

In [109]:
subset2 = df2[df2.model2 == 'Defect']
subset2

actual  model2
0    No Defect  Defect
1    No Defect  Defect
2    No Defect  Defect
3    No Defect  Defect
4    No Defect  Defect
..         ...     ...
183  No Defect  Defect
185  No Defect  Defect
193  No Defect  Defect
195  No Defect  Defect
198  No Defect  Defect

[90 rows x 2 columns]

In [110]:
model_precision = (subset2.actual == subset2.model2).mean()
print(f'model precision: {model_precision:.2%}')

model precision: 10.00%


**Model 3:**

In [111]:
subset3 = df3[df3.model3 == 'Defect']
subset3

actual  model3
1    No Defect  Defect
3    No Defect  Defect
5    No Defect  Defect
9    No Defect  Defect
13      Defect  Defect
..         ...     ...
193  No Defect  Defect
194     Defect  Defect
195  No Defect  Defect
198  No Defect  Defect
199  No Defect  Defect

[99 rows x 2 columns]

In [112]:
model_precision = (subset3.actual == subset3.model3).mean()
print(f'model precision: {model_precision:.2%}')

model precision: 13.13%


Since model 1 has the least false positives (least false positives), model 1 would be the best model to use to prevent unnecessary vactations to Hawaii.

### 4. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee). At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

### Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

In [113]:
df = pd.read_csv('gives_you_paws.csv')
df

actual model1 model2 model3 model4
0       cat    cat    dog    cat    dog
1       dog    dog    cat    cat    dog
2       dog    cat    cat    cat    dog
3       dog    dog    dog    cat    dog
4       cat    cat    cat    dog    dog
...     ...    ...    ...    ...    ...
4995    dog    dog    dog    dog    dog
4996    dog    dog    cat    cat    dog
4997    dog    cat    cat    dog    dog
4998    cat    cat    cat    cat    dog
4999    dog    dog    dog    dog    dog

[5000 rows x 5 columns]

In [114]:
df.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

Since dogs are the most common type of pet, we will create a baseline model of all 'dog'

In [115]:
df1 = df[['actual', 'model1']].copy()
df1['base_pred'] = 'dog'
df2 = df[['actual', 'model2']].copy()
df2['base_pred'] = 'dog'
df3 = df[['actual', 'model3']].copy()
df3['base_pred'] = 'dog'
df4 = df[['actual', 'model4']].copy()
df4['base_pred'] = 'dog'

#### a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

In [116]:
# Model 1

model1_accuracy = (df1.actual == df1.model1).mean()

print(f'   model1 accuracy: {model1_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model1 accuracy: 80.74%
baseline accuracy: 92.00%


In [117]:
# Model 2

model2_accuracy = (df2.actual == df2.model2).mean()

print(f'   model2 accuracy: {model2_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model2 accuracy: 63.04%
baseline accuracy: 92.00%


In [118]:
# Model 3

model3_accuracy = (df3.actual == df3.model3).mean()

print(f'   model3 accuracy: {model3_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model3 accuracy: 50.96%
baseline accuracy: 92.00%


In [119]:
# Model 4

model4_accuracy = (df4.actual == df4.model4).mean()

print(f'   model4 accuracy: {model4_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model4 accuracy: 74.26%
baseline accuracy: 92.00%


Models 1 and 4 show a higher accuracy than the baseline accuracy with model 1's being the highest overall at 80.74%.

#### b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recommend for Phase I? For Phase II?

    -Since we are looking for dog pictures, we will assign 'dog' to the positive case.

    -For Phase I, since the selected images could possibly be reviewed in Phase II, a model that is less likely to mislabel dogs as cats (False Negative) is preferable. A model with a higher recall metric will be the most appropriate.

Model 1:

In [120]:
subset1 = df1[df1.actual == 'dog']
subset1

# calculate recall for model
model1_recall = (subset1.actual == subset1.model1).mean()

print(f'model 1 recall: {model1_recall:.2%}')

# Precision: only look at the subset where we made positive prediction (i.e. prediction == 'dog')

subset1 = df1[df1.model1 == 'dog']
subset1
model1_precision = (subset1.actual == subset1.model1).mean()
print(f'model 1 precision: {model1_precision:.2%}')

model 1 recall: 80.33%
model 1 precision: 89.00%


Model 2:

In [126]:
subset2 = df2[df2.actual == 'dog']

# calculate recall for model
model2_recall = (subset2.actual == subset2.model2).mean()
print(f'model 2 recall: {model2_recall:.2%}')


subset2 = df2[df2.model2 == 'dog']
subset2
model2_precision = (subset2.actual == subset2.model2).mean()
print(f'model 2 precision: {model2_precision:.2%}')

model 2 recall: 49.08%
model 2 precision: 89.32%


Model 3:

In [128]:
subset3 = df3[df3.actual == 'dog']
subset3

model3_recall = (subset3.actual == subset3.model3).mean()
print(f'model 3 recall: {model3_recall:.2%}')


subset3 = df3[df3.model3 == 'dog']
subset3
model3_precision = (subset3.actual == subset3.model3).mean()
print(f'model 3 precision: {model3_precision:.2%}')

model 3 recall: 50.86%
model 3 precision: 65.99%


Model 4:

In [129]:
subset4 = df4[df4.actual == 'dog']
subset4

model4_recall = (subset4.actual == subset4.model4).mean()
print(f'model 4 recall: {model4_recall:.2%}')


subset4 = df4[df4.model4 == 'dog']
subset4
model4_precision = (subset4.actual == subset4.model4).mean()
print(f'model 4 precision: {model4_precision:.2%}')

model 4 recall: 95.57%
model 4 precision: 73.12%


Model 4 has the highest recall, so that would be the best model for Phase I.

For Phase II, we want the model that incorrectly tags cats as dogs (False Positive), so model 2 would be the best.

#### c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?

In [131]:
subset1 = df1[df1.actual == 'cat']
subset1

model1_recall = (subset1.actual == subset1.model1).mean()
print(f'model 1 recall: {model1_recall:.2%}')


subset1 = df1[df1.model1 == 'cat']
subset1
model1_precision = (subset1.actual == subset1.model1).mean()
print(f'model 1 precision: {model1_precision:.2%}')

model 1 recall: 81.50%
model 1 precision: 68.98%


In [132]:
subset2 = df2[df2.actual == 'cat']
subset2

model2_recall = (subset2.actual == subset2.model2).mean()
print(f'model 2 recall: {model2_recall:.2%}')


subset2 = df2[df2.model2 == 'cat']
subset2
model2_precision = (subset2.actual == subset2.model2).mean()
print(f'model 2 precision: {model2_precision:.2%}')

model 2 recall: 89.06%
model 2 precision: 48.41%


In [133]:
subset3 = df3[df3.actual == 'cat']
subset3

model3_recall = (subset3.actual == subset3.model3).mean()
print(f'model 3 recall: {model3_recall:.2%}')


subset3 = df3[df3.model3 == 'cat']
subset3
model3_precision = (subset3.actual == subset3.model3).mean()
print(f'model 3 precision: {model3_precision:.2%}')

model 3 recall: 51.15%
model 3 precision: 35.83%


In [134]:
subset4 = df4[df4.actual == 'cat']
subset4

model4_recall = (subset4.actual == subset4.model4).mean()
print(f'model 4 recall: {model4_recall:.2%}')


subset4 = df4[df4.model4 == 'cat']
subset4
model4_precision = (subset4.actual == subset4.model4).mean()
print(f'model 4 precision: {model4_precision:.2%}')

model 4 recall: 34.54%
model 4 precision: 80.72%


For Phase I, model 2 has the highest recall.

For Phase II, model 4 has the highest precision.

### Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

In [148]:
print('Model 1 accuracy =', sklearn.metrics.accuracy_score(df.actual, df1.model1, normalize=True))
print('Model 2 accuracy =', sklearn.metrics.accuracy_score(df.actual, df2.model2, normalize=True))
print('Model 3 accuracy =', sklearn.metrics.accuracy_score(df.actual, df3.model3, normalize=True))
print('Model 4 accuracy =', sklearn.metrics.accuracy_score(df.actual, df4.model4, normalize=True))

Model 1 accuracy = 0.8074
Model 2 accuracy = 0.6304
Model 3 accuracy = 0.5096
Model 4 accuracy = 0.7426


In [167]:
print('Model 1 precision =', round(sklearn.metrics.precision_score(df.actual, df.model1, pos_label='dog'), 4))
print('Model 2 precision =', round(sklearn.metrics.precision_score(df.actual, df.model2, pos_label='dog'), 4))
print('Model 3 precision =', round(sklearn.metrics.precision_score(df.actual, df.model3, pos_label='dog'), 4))
print('Model 4 precision =', round(sklearn.metrics.precision_score(df.actual, df.model4, pos_label='dog'), 4))

Model 1 precision = 0.89
Model 2 precision = 0.8932
Model 3 precision = 0.6599
Model 4 precision = 0.7312


In [168]:
print('Model 1 recall =', round(sklearn.metrics.recall_score(df.actual, df.model1, pos_label = 'dog'), 4))
print('Model 2 recall =', round(sklearn.metrics.recall_score(df.actual, df.model2, pos_label = 'dog'), 4))
print('Model 3 recall =', round(sklearn.metrics.recall_score(df.actual, df.model3, pos_label = 'dog'), 4))
print('Model 4 recall =', round(sklearn.metrics.recall_score(df.actual, df.model4, pos_label = 'dog'), 4))

Model 1 recall = 0.8033
Model 2 recall = 0.4908
Model 3 recall = 0.5086
Model 4 recall = 0.9557


In [184]:
print(sklearn.metrics.classification_report(df.actual, df.model1, target_names=['cat', 'dog']))

              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000



In [185]:
print(sklearn.metrics.classification_report(df.actual, df.model2, target_names=['cat', 'dog']))

              precision    recall  f1-score   support

         cat       0.48      0.89      0.63      1746
         dog       0.89      0.49      0.63      3254

    accuracy                           0.63      5000
   macro avg       0.69      0.69      0.63      5000
weighted avg       0.75      0.63      0.63      5000



In [186]:
print(sklearn.metrics.classification_report(df.actual, df.model3, target_names=['cat', 'dog']))

              precision    recall  f1-score   support

         cat       0.36      0.51      0.42      1746
         dog       0.66      0.51      0.57      3254

    accuracy                           0.51      5000
   macro avg       0.51      0.51      0.50      5000
weighted avg       0.55      0.51      0.52      5000



In [187]:
print(sklearn.metrics.classification_report(df.actual, df.model4, target_names=['cat', 'dog']))

              precision    recall  f1-score   support

         cat       0.81      0.35      0.48      1746
         dog       0.73      0.96      0.83      3254

    accuracy                           0.74      5000
   macro avg       0.77      0.65      0.66      5000
weighted avg       0.76      0.74      0.71      5000

